### ISSUE TITLE
Segfault when reading fill value of a vlen string HDF5 dataset

### ISSUE THREAD

**Issue by ajelenak** [h5py/h5py#1929]
Reading the fill value of a vlen string HDF5 dataset segfaults python with the error message:
```
python(63812,0x107808e00) malloc: *** error for object 0x7f9422c4d890: pointer being freed was not allocated
python(63812,0x107808e00) malloc: *** set a breakpoint in malloc_error_break to debug
zsh: abort      /opt/anaconda3/envs/h5py/bin/python
```
The `0x7f9422c4d890` object above is the value of the `buf_cstring0` variable in the `_conv.conv_vlen2str()` function: https://github.com/h5py/h5py/blob/a6c659037aea2d2cadaca893116c8ef0c6449fe7/h5py/_conv.pyx#L211. 

Reading values of a vlen string dataset seems to work.

Attached are a [sample HDF5 file](https://github.com/h5py/h5py/files/6812656/vlenstr-fillvalue.h5.zip) and a [Python example](https://github.com/h5py/h5py/files/6812661/test-vlen-str-fillvalue.py.zip) that causes segfaults. First it reads a vlen string element and then that dataset's fill value. The HDF5 file was created with the HDF5 library but I also have a file created with the netCDF/HDF5 library combo that causes segfaults.

---

Summary of the h5py configuration:
```
h5py    master branch
HDF5    1.12.1
Python  3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 05:02:20) 
[Clang 11.0.1 ]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.20.3
cython (built with) 0.29.23
numpy (built against) 1.20.3
HDF5 (built against) 1.12.1
```

**Comment by takluyver** [comment]
Do you have to do anything special to expose this? Your example runs OK for me (on Linux):

```
Dataset /with-fillvalue string datatype = string_info(encoding='ascii', length=None)
Reading dataset first element
dset[0] = b'Trouble'
Finished reading dset[0]
Reading dataset fill value
Fill value: b'...'
Dataset /empty-fillvalue string datatype = string_info(encoding='ascii', length=None)
Reading dataset first element
dset[0] = b'Trouble'
Finished reading dset[0]
Reading dataset fill value
Fill value: b''
The End
```

<details>
<summary>Script from attachment, for quicker reference</summary>

```python
import h5py

with h5py.File('vlenstr-fillvalue.h5', mode='r') as h5f:
    for name in ('with-fillvalue', 'empty-fillvalue'):
        dset = h5f[name]
        str_dt = h5py.check_string_dtype(dset.dtype)
        if str_dt:
            print(f'Dataset {dset.name} string datatype = {str_dt}')
            print('Reading dataset first element')
            print(f'dset[0] = {dset[0]}')
            print('Finished reading dset[0]')
            print('Reading dataset fill value')
            print(f'Fill value: {dset.fillvalue}')
    print('The End')
```
</details>

**Comment by ajelenak** [comment]
Fixed in #2044.
